pyro/optim/multi.py

```diff
@@ -0,0 +1,138 @@
+
+import torch
+
+from pyro.ops.newton import newton_step_2d
+from pyro.optim.optim import PyroOptim
+
+
+class MultiOptimizer(object):
+    """
+    Base class of optimizers that make use of higher-order derivatives.
+
+    Higher-order optimizers generally use :func:`torch.autograd.grad` rather
+    than :meth:`torch.Tensor.backward`, and therefore require a different
+    interface from usual Pyro and PyTorch optimizers. In this interface,
+    the :meth:`step` method inputs a ``loss`` tensor to be differentiated,
+    and backpropagation is triggered one or more times inside the optimizer.
+
+    Derived classes must implement :meth:`step` to compute derivatives and
+    update parameters in-place.
+
+    Example::
+
+        tr = poutine.trace(model).get_trace(*args, **kwargs)
+        loss = -tr.log_prob()
+        params = {name: pyro.param(name).unconstrained()
+                  for name in pyro.get_param_store().get_all_param_names()}
+        optim.step(loss, params)
+    """
+    def step(self, loss, params):
+        """
+        Performs an in-place optimization step on parameters given a
+        differentiable ``loss`` tensor.
+
+        Note that this detaches the updated tensors.
+
+        :param torch.Tensor loss: A differentiable tensor to be minimized.
+            Some optimizers require this to be differentiable multiple
+            times.
+        :param dict params: A dictionary mapping param name to unconstrained
+            value as stored in the param store.
+        """
+        updated_values = self.get_step(loss, params)
+        for name, value in params.items():
+            with torch.no_grad():
+                value[...] = updated_values[name]
+
+    def get_step(self, loss, params):
+        """
+        Computes an optimization step of parameters given a differentiable
+        ``loss`` tensor, returning the updated values.
+
+        Note that this preserves derivatives on the updated tensors.
+
+        :param torch.Tensor loss: A differentiable tensor to be minimized.
+            Some optimizers require this to be differentiable multiple
+            times.
+        :param dict params: A dictionary mapping param name to unconstrained
+            value as stored in the param store.
+        :return: A dictionary mapping param name to updated unconstrained
+            value.
+        :rtype: dict
+        """
+        raise NotImplementedError
+
+
+class PyroMultiOptimizer(MultiOptimizer):
+    """
+    Facade to wrap :class:`~pyro.optim.optim.PyroOptim` objects
+    in a :class:`MultiOptimizer` interface.
+    """
+    def __init__(self, optim):
+        self.optim = optim
+
+    def step(self, loss, params):
+        values = params.values()
+        grads = torch.autograd.grad(loss, values, create_graph=True)
+        for x, g in zip(values, grads):
+            x.grad = g
+        self.optim(values)
+
+
+class TorchMultiOptimizer(PyroMultiOptimizer):
```

Comment:
what is the purpose of torch multi optimizer since it is essentially a pyro multi optimizer (`optim` is wrapped in `PyroOptim` during initialization anyway)?  is there a use case for this?

Reply:
This is intended to help us explore the possibility of moving towards `grad`-based optimizers rather than `backward`-based `PyroOptim` #628 . @eb8680 and I want to use this assess whether we can get by with the `TorchMultiOptimizer` interface and possibly deprecate the `PyroOptim` interface, moving `PyroOptim`'s logic into `TorchMultiOptimizer`. We do not intend to address whether we should make this refactoring, only whether it is possible.